pkg/controller/awsfederatedaccountaccess/awsfederatedaccountaccess_controller_test.go

```diff
@@ -0,0 +1,422 @@
+
+	awsOutputRole := &iam.CreateRoleOutput{
+		Role: &iam.Role{},
+	}
+
+	tests := []struct {
+		name                string
+		uidLabel            map[string]string
+		createIAMRoleOutput *iam.Role
+		expectedErr         error
+	}{
+		{
+			name:                "Test for UID label present",
+			uidLabel:            map[string]string{"uid": "abcd"},
+			createIAMRoleOutput: awsOutputRole.Role,
+			expectedErr:         nil,
+		},
+		{
+			name:                "Test for UID label missing",
+			uidLabel:            map[string]string{},
+			createIAMRoleOutput: nil,
+			expectedErr:         errors.New("Failed to get UID label"),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+
+			mocks := setupDefaultMocks(t, []runtime.Object{})
+
+			mockAWSClient := mock.NewMockClient(mocks.mockCtrl)
+
+			defer mocks.mockCtrl.Finish()
+
+			if test.createIAMRoleOutput != nil {
+				mockAWSClient.EXPECT().CreateRole(gomock.Any()).Return(
+					awsOutputRole,
+					nil,
+				)
+			}
+
+			r := ReconcileAWSFederatedAccountAccess{}
+
+			afr := awsv1alpha1.AWSFederatedRole{
+				ObjectMeta: v1.ObjectMeta{
+					Name: "",
+				},
+				Spec: awsv1alpha1.AWSFederatedRoleSpec{
+					AWSCustomPolicy: newTestAwsCustomPolicyBuilder().awsCustomPol},
+			}
+
+			afaa := awsv1alpha1.AWSFederatedAccountAccess{
+				ObjectMeta: v1.ObjectMeta{
+					Labels: test.uidLabel},
+				Spec: awsv1alpha1.AWSFederatedAccountAccessSpec{
+					ExternalCustomerAWSIAMARN: "",
+				},
+			}
+
+			createRoleOutput, err := r.createIAMRole(mockAWSClient, afr, afaa)
+			assert.Equal(t, err, test.expectedErr)
+			assert.Equal(t, createRoleOutput, test.createIAMRoleOutput)
+		})
+	}
+}
+
+func TestCreateOrUpdateIAMPolicy(t *testing.T) {
+
+	mocks := setupDefaultMocks(t, []runtime.Object{})
+
+	mockAWSClient := mock.NewMockClient(mocks.mockCtrl)
+
+	defer mocks.mockCtrl.Finish()
+
+	afr := awsv1alpha1.AWSFederatedRole{
+		Spec: awsv1alpha1.AWSFederatedRoleSpec{
+			AWSCustomPolicy: newTestAwsCustomPolicyBuilder().awsCustomPol},
+	}
+
+	afaa := awsv1alpha1.AWSFederatedAccountAccess{
+		ObjectMeta: v1.ObjectMeta{
+			Labels: map[string]string{"uid": ""}},
+	}
+
+	uidLabel := afaa.Labels["uid"]
+	policyName := afr.Spec.AWSCustomPolicy.Name + "-" + uidLabel
+
+	type awsStatement struct {
+		Effect    string                 `json:"Effect"`
+		Action    []string               `json:"Action"`
+		Resource  []string               `json:"Resource,omitempty"`
+		Condition *awsv1alpha1.Condition `json:"Condition,omitempty"`
+		Principal *awsv1alpha1.Principal `json:"Principal,omitempty"`
+	}
+
+	statements := []awsStatement{}
+	for _, sm := range afr.Spec.AWSCustomPolicy.Statements {
+		var a awsStatement = awsStatement(sm)
+		statements = append(statements, a)
+	}
+
+	policyDoc := struct {
+		Version   string
+		Statement []awsStatement
+	}{
+		Version:   "2012-10-17",
+		Statement: statements,
+	}
+
+	jsonPolicyDoc, err := json.Marshal(&policyDoc)
+	assert.Nil(t, err)
```

Comment:
Don't think this needs to be asserted either as it's static data, we're not randomly seeding anything so there's no speculation for data variance